### ISSUE TITLE
Selecting the appropriate JRuby JSR223 Context, Shared Variable, and local variable settings

### ISSUE THREAD

**Issue by boc-tothefuture** [boc-tothefuture/openhab-jruby#46]
[OpenHAB](https://www.openhab.org/) is an open source Java based home automation built on top of Apache Karaf and utilizing OSGI. 

OpenHAB supports JSR223 to enable scripting languages to create automation rules.  There are several JSR223 implementations with python being the most popular currently.  This project aims to add Ruby as one of the options for creating automation rules within OpenHAB.  In particular, this project aims to create an intuitive easy to use rules language and is influenced by the design of Chef.  More details about project goals are available [here](https://github.com/boc-tothefuture/openhab-jruby/blob/main/README.md).

This issue is created to determine the best selection of JRuby/JSR223 settings to enable the goals of this project. 

In general, rules are structured as follows:

```ruby
require 'openhab'

rule 'name' do
   <zero or more triggers>
   <zero or more execution blocks>
   <zero or more guards>
end
```

Many examples of rules are available  [here](https://github.com/boc-tothefuture/openhab-jruby/blob/main/README.md).

OpenHAB JSR223 rule processing and execution structure. 

JSR223 rules files are placed in a specific subdirectory that OpenHAB scans on starts and also watches for changes.  If any changes occur, OpenHAB requests the ScriptEngine reevaluate the script.  The code for script evaluation is this:

```java
@Override
    public void loadScript(String engineIdentifier, InputStreamReader scriptData) {
        ScriptEngineContainer container = loadedScriptEngineInstances.get(engineIdentifier);
        if (container == null) {
            logger.error("Could not load script, as no ScriptEngine has been created");
        } else {
            ScriptEngine engine = container.getScriptEngine();
            try {
                engine.eval(scriptData);
                if (engine instanceof Invocable) {
                    Invocable inv = (Invocable) engine;
                    try {
                        inv.invokeFunction("scriptLoaded", engineIdentifier);
                    } catch (NoSuchMethodException e) {
                        logger.trace("scriptLoaded() is not defined in the script: {}", engineIdentifier);
                    }
                } else {
                    logger.trace("ScriptEngine does not support Invocable interface");
                }
            } catch (Exception ex) {
                logger.error("Error during evaluation of script '{}': {}", engineIdentifier, ex.getMessage());
            }
        }
    }
```
Rules files are loaded in-order by a single thread, [based on name and directory structure](https://www.openhab.org/docs/configuration/jsr223.html#script-locations).

The directory watching trigger has a synchronization point on loading and further evaluating scripts, such that only a single script file will be evaluated at any one time. 

The evaluation process, results in the creation of Java objects which are called "Rules". These rules have a set of one or more triggers which the OpenHab engine is watching and if they meet the conditions of a trigger, they will be invoked. The invocation process comes back into the JSR223 engine to execute the actions specified by the rule.  In the high level example above, I refer to those as execution blocks.  The framework assigns each rule is assigned its own thread (for the lifetime of the rule existing). 

By way of example, to highlight the threading model:
If we have 5 files ending in '.rb', the OpenHAB engine will, in serial call the above `loadScript` method for each file. If each file contains 4 rules each after evaluation 20 threads will be created and 20 threads *could* execute simultaneously processing the various execution blocks. 

Expected behavior by end users:

- Users expect to segment their rules into rules files based on personal criteria such as functional (all things pertaining to lights) or thing-based (all my rules related to Alexa are in alexa.rb, etc).  
- Minimal setup and/or fuss.  Some people have a programming background, others do not.  Creating complex objects before starting any set of rules would not be well received.  Easy things should be easy, difficult things should be possible.
- What happens in a rules file, stays in a rules file unless instructed otherwise. In other words, local and instance variables should not bleed into other rules files, globals could and should.  
- Users expect (at a minimum) to be able to share information between rules that exist in the same file (which are being executed by different threads). This is a highly common scenario.

Rule creation uses instance_exec to aid in the creation of the DSL, which sets the context during rule creation to operate within the `rule 'name' do` block.  In a branch (not released) there is a change such that the execution blocks are evaluated within the context of whomever create the rule.  This permits the sharing of instance variables between rules. 
To limit the scope of that sharing, I have introduced the concept of a 'rule_set' which is a block in which rules are placed that need to share instance variables. 

For example:
```ruby

rule_set do
  rule 'one' do
    run { @foo = 'bar' }
 end

 rule 'two' do
   run { logger.info("@foo is #{@foo}")
 end

end
```
After proposing this, the community asked if rule_sets could be named so they [could be referenced by name](https://community.openhab.org/t/jruby-openhab-rules-system/110598/99?u=broconne) and utilized in other rules files.  I attempted to implement this by placing the rulesets into a Java Concurrent HashMap maintained by the OpenHab engine but because the various JRuby runtimes creating different instances of the String class, lookup into that shared runtime failed. I would say this request is probably a "nice to have" rather than a must have and could be solved by other methods, such as global variables. 

The desired behavior for ruby variables is the following:

| variable type | desired scope | 
|--------------| -------------- |
| local               | block (rule) or current file if outside block |
| instance        | current file - not visible from other files |
| global            | visible from all files and rules |

The [cucumber tests in this branch](https://github.com/boc-tothefuture/openhab-jruby/blob/variable_scope/features/ruby.feature), demonstrate the expected scoping rules. 

The current JSR223 settings for JRuby are as follows
```java
static {
        System.setProperty("org.jruby.embed.sharing.variables", "false");
        System.setProperty("org.jruby.embed.localcontext.scope", "threadsafe");
        System.setProperty("org.jruby.embed.localvariable.behavior", "transient");
    }
```

With those settings, local and instance variables appear to work as desired.  However, global and named rule_sets are not visible outside the current file.

**Comment by headius** [comment]
I sent some messages on Matrix and then realized it would probably be better to do them here.

First off there is no standard way to isolate instance variables in the same objects across different files. You must use a different object or a different engine/container to isolate that state.

However the general mechanism for doing this in DSLs is exactly what you are already using in some capacity: `instance_exec`.

In your example DSL you have a `rule_set` containing two `rule` each with a `run`. The `run` block is what you would want to isolate across files or invocations by creating a new object and using `instance_exec` to execute that block.

This is similar to how rspec uses `instance_exec` to isolate a spec's instance variables from the top-level global "main" object.

The threadsafe container mode you attempted to use led to an issue with multiple of the same string key being inserted into a shared global hash. As we discussed on Matrix, this is because of how the threadsafe mode works: it provides a complete isolated engine/container per thread. Your code tripped up this isolation by sharing a single Java Map containing the shared Ruby hash object, and the keys inserted into it come from different containers so they look like different objects even though they have the same string content.

I believe the two modes you want to look at are `SINGLETHREAD` and `CONCURRENT`.

The `SINGLETHREAD` mode is poorly named, and basically means there will be no thread isolation or synchronization. A single JRuby runtime is used and threads may potentially see each others updates to top-level local variables and global variables. You would be responsible for isolating different files or threads, probably by using `instance_exec` to run sensitive code against different objects.

The `CONCURRENT` mode is similar to `SINGLETHREAD` but uses thread-local stores for those top-level local variables (and possibly globals, I am unsure at the moment).

**Comment by jimtng** [comment]
Currently using the main branch, I am able to use class variables in the same rule file.

```ruby
@@myvar = 'foo'

rule 'rule1' do
  on_start
  run do
    logger.info("rule #{name} @@myvar = #{@@myvar}") 
    @@myvar = 'baz'
  end
end

rule 'rule2' do
  on_start
  run do
    logger.info("rule #{name} @@myvar = #{@@myvar}") 
  end
end
```

Output:
```
14:16:09.373 [INFO ] [.OpenHAB::Core::DSL::Rule::RuleConfig] - rule rule1 @@myvar = foo
14:16:09.375 [INFO ] [.OpenHAB::Core::DSL::Rule::RuleConfig] - rule rule2 @@myvar = baz
```

**Comment by boc-tothefuture** [comment]
@headius  -  Thank you for the reply.. I apologize for the delay in my response.  We had a build go out with a bunch of regressions and needed to spend time fixing the build process to ensure testing.

I think if we require the user to always put a rules inside of a rule_set we are going to have usability issues as this is different than any of the other scripting languages. I was attempting to use the rule_set just for places where we want to share instance variables.  If we end up in a situation where top-level local variables are shared by default between files we are going to run into a situation where users are hitting a lot of weird conditions.

I think it would help me if we visualized this in a table. I phrased it in terms of files/threads because that is the user experience.  I *think* that when I say per file, that translates to does JRuby provide different instances of the base Object. 

Can you tell me if the following is correct?


| Variable Type      | Singleton                   | ThreadSafe                                           | SingleThread                 | Concurrent                   |
| ------------------ | --------------------------- | ---------------------------------------------------- | ---------------------------- | ---------------------------- |
| top-level local    | Shared by all files/therads | Isolated per file                                    | Shared by all files/threads? | Shared by all files/threads? |
| top-level instance | Shared by all files/threads | Isolated per thread, inaccessible across files       | Shared by all files/threads? | Shared by all files/threads? |
| Global             | Shared by all files/threads | Isolated per file/threads, inaccessible across files | Shared by all files/threads? | Shared by all files/threads? |

**Comment by headius** [comment]
I think this is right except for the instance vars + threadsafe... from what I have seen there is no mode that will either isolate instance vars across files or provide a different top-level anonymous object for top-level instance variables. I think it would be easy to prove this one way or another if you already have code working, right?

**Comment by headius** [comment]
I want to ping @jdsundberg here because his team built a product with embedded JRuby for scripting and he or someone from his team might be able to offer some perspective on how they did embedded scripting in a multi-user environment.

**Comment by bchristenson** [comment]
I work with @jdsundberg and can speak a bit about how we are using Embedded JRuby in our platform here at https://kineticdata.com.

Our workflow automation component embeds JRuby into a Java web application and allows users to import workflow "handlers" (zip files containing input/output definitions, the actual Ruby logic and library dependencies).  Our use case is a bit closer to importing Ruby classes into a single shared namespace (per tenant, which we separate with independent web applications), but we do have some other aspects of the platform that leverage scoped evaluation or DSLs.

I'm not sure if this is close to what you are looking for, but I've included a simplified snippet of a pattern that we have used to share instance scope in other areas.  In this example it is possible for a rule to be run from two different threads, and since the container is SINGLETHREAD, it is up to you to manually enforce whatever thread safety is appropriate.

```
package com.kineticdata.samples;

import org.jruby.embed.LocalContextScope;
import org.jruby.embed.ScriptingContainer;

public class SampleDriver {
    
    public static void main(String[] args) {
        // Initialize a container
        ScriptingContainer container = new ScriptingContainer(LocalContextScope.SINGLETHREAD);

        // Simulate setting a global
        container.runScriptlet(
            "$global = 'GLOBAL'");
        // Simulate defining basic DSL
        container.runScriptlet(
            "$RULESETS = {}");
        container.runScriptlet(
            "$RULES = {}");
        container.runScriptlet(
            "class Rule\n" +
            "  attr_accessor :ruleset\n" +
            "  attr_accessor :handler\n" +
            "  def initialize(ruleset, handler)\n" +
            "    @ruleset = ruleset\n" +
            "    @handler = handler\n" +
            "  end\n" +
            "end");
        container.runScriptlet(
            "def rule(name, ruleset_name=nil, &handler)\n" + 
            "  ruleset = if ruleset_name\n" + 
            "    $RULESETS[ruleset_name] ||= Object.new\n" + 
            "  else\n" + 
            "    Object.new\n" + 
            "  end\n" + 
            "  $RULES[name] = Rule.new(ruleset, handler)\n" + 
            "end");
        container.runScriptlet(
            "def call(name, params={})\n" + 
            "  rule = $RULES[name]\n" + 
            "  rule.ruleset.instance_exec(params, &rule.handler)\n" + 
            "end");
        
        // Define 3 rules (1 with independent scopes, 2 with shared)
        container.runScriptlet(
            "rule 'A' do |params|\n" +
            "  local = local || %|LOCAL-#{params[:val]}|\n" +
            "  @instance = @instance || %|INSTANCE-#{params[:val]}|\n" +
            "  %|#{$global} #{@instance} #{local}|\n" +
            "end");
        container.runScriptlet(
            "rule 'B1', 'Shared' do |params|\n" +
            "  local = local || 'LOCAL-B1'\n" +
            "  @instance = @instance || 'INSTANCE-B1'\n" +
            "  %|#{$global} #{@instance} #{local}|\n" +
            "end");
        container.runScriptlet(
            "rule 'B2', 'Shared' do |params|\n" +
            "  local = local || 'LOCAL-B2'\n" +
            "  @instance = @instance || 'INSTANCE-B2'\n" +
            "  %|#{$global} #{@instance} #{local}|\n" +
            "end");
        
        // Simulate invoking the 4 rules
        System.out.println(container.runScriptlet("call 'A', {:val => 'A1'}"));
        System.out.println(container.runScriptlet("call 'A', {:val => 'A2'}"));
        System.out.println(container.runScriptlet("call 'B1'"));
        System.out.println(container.runScriptlet("call 'B2'"));
        
        // GLOBAL INSTANCE-A1 LOCAL-A1
        // GLOBAL INSTANCE-A1 LOCAL-A2
        // GLOBAL INSTANCE-B1 LOCAL-B1
        // GLOBAL INSTANCE-B1 LOCAL-B2
    }   
}
```

**Comment by boc-tothefuture** [comment]
@ccutrer  did you look at this issue yet?  Your ruby background might be helpful as we determine if this is settled or not.

**Comment by ccutrer** [comment]
so as a rubyist, these are things I would expect in terms of namespacing:

 * locals from one file definitely do not leak to other files
 * top level methods, constants, and globals from one file are available from other files _only_ if I require the file
 * if I do require the same file from multiple files, I expect those top level items to refer to the _same_ instance regardless of which file I'm accessing from.

1, and 2 are currently true. 3 is not: each rules script gets its own environment, even if you `require` the same file, the globals are their own copy. this even applies to `require "openhab"`, and is why it needs to be done from each script. IF this changes so that multiple script files share an execution environment, it would be nice that the openhab-scripting library won't need to re-set up when the script is loaded, and will then satisfy number 3, but we would need to be careful to ensure 1 and 2 are still true. However, I'm wary of doing it. Even though 3 is not currently my initial expectations, it makes sense after some experience with jruby rules, and seems like it would be fraught with problems to get it acting "correctly". For example, if 3 becomes true, and you set a timer in a script file and store it in a top-level, local variable hash (as I often do), but then change the file, the old instance of the hash is lost, but the timer would live on, with no way to access it anymore. And when it triggers, it will call the old version of the code that doesn't exist anymore. Talk about confusing a casual user. The only way to get it to reload would be to restart the jrubyscripting bundle (or more likely a more casual user, would restart all of openhab). My day job is working with a giant Rails app, and professional ruby developers still get frustrated and confused when old code that doesn't exist anymore is still running (due to the size of the app, it's encouraged to use [spring](https://github.com/rails/spring), which only reloads piecemeal on demand, and isn't always perfect).

So... I'd say that unless someone comes along and is adamant that they need shared state among different rules files, and is wiling to put in the work to do it correctly, the current state of things is best for casual users, and workable for advanced users.

**Comment by boc-tothefuture** [comment]
Reopening to keep this all in a single place.  
In https://github.com/boc-tothefuture/openhab2-addons/pull/9 
@jimtng said:

> > I am running singlethread on my production environment as of today. So far everything seems to be running as normal.
> 
> I don't fully understand the issue surrounding multi-threads with the jruby. Could you give an example scenario where singlethread would cause a problem?

> If I understand it correctly:
> 
> Each rule file gets its own scriptengine instance
> rules in different files should have no problems because they are running in a completely different script engine, like a separate 'virtual machine'
> multiple rules in the same file will share the same scriptengine, and can access the same instance variables and top-level variables. However they could be running concurrently through different threads, so modifications to variables that are visible to multiple rules (within the same file) are subject to multi-threading issues and therefore may need to be synchronised
> threadsafe option automatically manage the access to such variables, therefore no explicit synchronisations are required
> singlethread doesn't have such management
> I could probably gain a better understanding if I spent some time digging into jruby's code but that would take time and some serious brain exercise.

**Comment by boc-tothefuture** [comment]
My understanding was that SingleThreaded crossed ScriptEngines.. if that is not the case then I would think SingleThreaded would be ok.

**Comment by boc-tothefuture** [comment]
Looking again at https://github.com/jruby/jruby/wiki/RedBridge#Context_Type it does look like single threaded should work in most cases because rules run in their own threads users would need to do synchronization for any shared objects between threads.
